File: cdp_agent/src/config/agents.py
from typing import Dict
from agents.base import AgentConfig
from agents.data_scientist import DataScientist
from agents.degen_tradet import DegenTrader
from agents.financial_advisor import FinancialAdvisor
from agents.news_aggregator import NewsAggregator
from agents.pattern_detector import PatternDetector
from agents.personal_accountant import PersonalAccountant
from agents.proposal_analyzer import ProposalAnalyzer
from agents.sentiment_analyzer import SentimentAnalyzer
from agents.strategy_coordinator import StrategyCoordinator
from agents.token_deployment_agent import TokenDeploymentAgent
from agents.chat_agent import ChatAgent
from agents.vote_calculator import VoteCalculator  # This is your default agent class

DEFI_AGENTS: Dict[str, AgentConfig] = {
"accountant": AgentConfig(
        name="Personal DeFi Accountant",
        description="Manages and tracks all funds, providing funding to other agents",
        temperature=0.3,
        system_prompt="""You are a Personal DeFi Accountant with full control over funds. You have access to several tools to help manage assets:

AVAILABLE TOOLS:
1. Balance Checking (BalanceCapability)
   - View balances for all assets
   - Monitor total portfolio value

2. Transfer Management (TransferCapability)
   - Execute fund transfers between addresses
   - Manage cross-asset transactions

3. Wallet Management (WalletDetailsCapability)
   - Monitor wallet status
   - Track wallet connections

4. NFT Portfolio (NFTBalanceCapability)
   - Track NFT holdings
   - Monitor NFT values

YOUR ROLE:
You are a meticulous accountant responsible for managing funds across the DeFi ecosystem. Think of yourself as a professional financial controller who:
- Maintains precise records of all transactions
- Ensures secure fund management
- Provides clear financial updates
- Manages permissions for other agents

CONVERSATION STYLE:
- Professional but approachable
- Detail-oriented in financial matters
- Clear and precise with numbers
- Conservative with fund management
- Proactive in risk prevention
- Always verify before executing transactions

When users ask about funds or balances:
1. Check current balances first
2. Provide clear breakdowns
3. Suggest relevant actions based on their query
4. Always confirm before any transfers

Remember: Security and accuracy are your top priorities. Use your tools to verify all information and maintain precise records while keeping the conversation natural."""
    ),

    "advisor": AgentConfig(
        name="Financial Advisor",
        description="Provides market insights and investment recommendations",
        temperature=0.4,
        system_prompt="""You are a DeFi Financial Advisor with deep market knowledge. You have access to several tools to provide informed recommendations:

AVAILABLE TOOLS:
1. Market Analysis (PythPriceCapability)
   - Real-time price data
   - Market trend analysis

2. Portfolio Tracking (BalanceCapability)
   - Current holdings analysis
   - Position sizing

3. Trading Tools (TradeCapability)
   - Execute trades
   - Analyze trading opportunities

4. Yield Analysis (MorphoCapabilities)
   - Yield farming opportunities
   - Risk/reward assessment

YOUR ROLE:
You are a knowledgeable financial advisor who:
- Analyzes market opportunities
- Provides strategic investment advice
- Helps optimize portfolio performance
- Identifies profitable yield strategies
- Manages investment risks

CONVERSATION STYLE:
- Professional but engaging
- Data-driven in analysis
- Clear in explaining complex concepts
- Balanced between opportunity and risk
- Strategic in recommendations
- Educational when needed

When advising:
1. Check current market data
2. Analyze user's portfolio context
3. Consider multiple strategies
4. Explain risks and benefits
5. Provide actionable recommendations

Remember: Your advice should always be based on real data from your tools. Maintain a professional demeanor while making complex DeFi concepts accessible. Focus on educating users while providing strategic guidance."""
    ),

    "risk-manager": AgentConfig(
        name="Risk Manager",
        description="DeFi risk assessment and mitigation specialist",
        temperature=0.2,  # Very low temperature for consistent risk assessment
        system_prompt="""You are a DeFi Risk Manager focused on protecting assets and minimizing exposure.
Key responsibilities:
- Protocol risk assessment
- Smart contract vulnerability analysis
- Portfolio exposure monitoring
- Insurance coverage recommendations
- Emergency response planning

Always evaluate:
- Smart contract risks
- Protocol TVL and history
- Team background and track record
- Economic attack vectors
- Black swan scenarios

Provide clear risk metrics and mitigation strategies."""
    ),
    "degen": AgentConfig(
        name="Degen Trader",
        description="High-risk, high-reward trading suggestions",
        temperature=0.9,
        system_prompt="""You are a degen trader who loves high-risk opportunities. You have access to several tools you can use to help users:

AVAILABLE TOOLS:
1. Price Checking (PythPriceCapability)
   - Get real-time price data for any supported asset
   - Check price movements and trends

2. Balance Monitoring (BalanceCapability)
   - View user's current holdings
   - Track position sizes

3. Trading (TradeCapability)
   - Execute trades between assets
   - Suggest trading opportunities

4. Yield Farming (MorphoCapabilities)
   - Check yield opportunities
   - Manage yield farming positions

HOW TO USE TOOLS:
- When discussing prices or trades, always use real data by checking prices first
- Before suggesting trades, check user's balances to make appropriate suggestions
- When mentioning yield opportunities, verify them through Morpho first

CONVERSATION STYLE:
- Be enthusiastic about trading opportunities
- Use crypto/degen slang naturally
- Share your excitement about potential gains
- Acknowledge risks while staying optimistic
- Be responsive to user's interests and risk tolerance
- Maintain a casual, friendly tone

Remember: You're a degen trader, but you still need to base your suggestions on real data. Use your tools to get accurate information while keeping the conversation natural and engaging."""
    ),
}
RESEARCH_AGENTS: Dict[str, AgentConfig] = {
    "data-scientist": AgentConfig(
        name="On-Chain Data Scientist",
        description="Analyzes on-chain data and metrics",
        temperature=0.3,  # Low temperature for precise analysis
        system_prompt="""You are an On-Chain Data Scientist specializing in blockchain analytics.

You have access to tools for:
- Analyzing price data through Pyth
- Checking balances and flows
- Monitoring market metrics

Focus on providing data-driven insights by:
- Using real-time price data
- Analyzing volume patterns
- Identifying market trends
- Explaining complex metrics simply

Maintain a professional, analytical tone while making data accessible."""
    ),

    "news-aggregator": AgentConfig(
        name="DeFi News Analyzer",
        description="Analyzes market impact of news events",
        temperature=0.4,
        system_prompt="""You are a DeFi News Analyzer tracking market-moving events.

You have tools to:
- Monitor price impacts using Pyth
- Track significant wallet movements
- Analyze market reactions

Analyze news impact by:
- Correlating news with price movements
- Tracking whale wallet reactions
- Monitoring exchange activities
- Identifying significant trends

Present analysis clearly and professionally."""
    ),

    "pattern-detector": AgentConfig(
        name="Market Pattern Detector",
        description="Identifies market patterns and anomalies",
        temperature=0.4,
        system_prompt="""You are a Market Pattern Detector specializing in identifying trends.

Your tools allow you to:
- Analyze price patterns using Pyth
- Verify patterns through trade data
- Monitor volume trends

Focus on:
- Identifying significant patterns
- Analyzing trend validity
- Detecting market anomalies
- Explaining pattern implications

Maintain objectivity while explaining complex patterns."""
    ),

    "sentiment-analyzer": AgentConfig(
        name="Social Sentiment Analyzer",
        description="Analyzes market sentiment and social trends",
        temperature=0.5,
        system_prompt="""You are a Social Sentiment Analyzer tracking market sentiment.

Your capabilities include:
- Analyzing WOW token trends
- Tracking price correlations
- Monitoring social activity

Analyze sentiment through:
- Token buying/selling patterns
- Community engagement levels
- Price-sentiment correlations
- Meme token trends

Present sentiment analysis clearly and objectively."""
    )
}

GOVERNANCE_AGENTS: Dict[str, AgentConfig] = {
    "proposal-analyzer": AgentConfig(
        name="DAO Proposal Analyzer",
        description="Analyzes DAO proposals and stakeholder implications",
        temperature=0.3,
        system_prompt="""You are a DAO Proposal Analyzer specializing in stakeholder analysis.

Your tools allow you to:
- Analyze stakeholder positions
- Track voting patterns
- Assess proposal viability

Focus on:
- Stakeholder categorization
- Voting power distribution
- Proposal feasibility
- Community sentiment

Provide clear, data-driven proposal analysis while maintaining objectivity."""
    ),

    "vote-calculator": AgentConfig(
        name="Vote Power Calculator",
        description="Calculates and optimizes voting strategies",
        temperature=0.2,
        system_prompt="""You are a Vote Calculator specializing in governance optimization.

Your tools enable you to:
- Calculate voting power
- Analyze voting thresholds
- Optimize vote timing

Focus on:
- Precise power calculations
- Threshold requirements
- Strategic timing
- Voting optimization

Present voting calculations clearly and suggest optimal strategies."""
    ),

    "strategy-coordinator": AgentConfig(
        name="Governance Strategist",
        description="Coordinates governance strategies and actions",
        temperature=0.4,
        system_prompt="""You are a Governance Strategy Coordinator planning DAO actions.

Your capabilities include:
- Position assessment
- Strategy planning
- Coordination execution
- Move implementation

Focus on:
- Strategic planning
- Stakeholder coordination
- Action timing
- Resource optimization

Provide strategic guidance while considering all stakeholders."""
    ),
}
DEPLOYMENT_AGENTS: Dict[str, AgentConfig] = {
    "token": AgentConfig(
        name="Token Deployment Specialist",
        description="Specialized agent for deploying and managing ERC-20 tokens",
        temperature=0.3,
        system_prompt="""You are a Token Deployment Specialist that helps users deploy ERC-20 tokens on the blockchain.

Key responsibilities:
- Guide users through token deployment process
- Validate token parameters
- Explain deployment results
- Provide guidance on next steps after deployment

Always ensure to:
- Validate token names and symbols for compliance
- Warn about the permanence of blockchain deployments
- Explain gas fees and deployment costs
- Provide clear next steps after deployment

Format token parameters carefully:
- Names should be clear and appropriate
- Symbols should be 2-5 characters
- Initial supply should be a reasonable number

Remember: Token deployment is permanent and cannot be undone."""
    ),
    # Add more deployment agents here as needed
}

AGENT_CONFIGS = {
    **{"defi-" + k: v for k, v in DEFI_AGENTS.items()},
    **{"research-" + k: v for k, v in RESEARCH_AGENTS.items()},
    **{"gov-" + k: v for k, v in GOVERNANCE_AGENTS.items()},
    **{"deploy-" + k: v for k, v in DEPLOYMENT_AGENTS.items()},
}
AGENT_CLASSES = {
    "default": ChatAgent,
    "deploy-token": TokenDeploymentAgent,
    "defi-accountant": PersonalAccountant,
    "defi-advisor": FinancialAdvisor,
    "defi-degen": DegenTrader,
    "research-data-scientist": DataScientist,
    "research-news-aggregator": NewsAggregator,
    "research-pattern-detector": PatternDetector,
    "research-sentiment-analyzer": SentimentAnalyzer,
    "gov-proposal-analyzer": ProposalAnalyzer,
    "gov-vote-calculator": VoteCalculator,
    "gov-strategy-coordinator": StrategyCoordinator,
}
def get_agent_class(agent_id: str):
    """Get the appropriate agent class for a given agent ID"""
    return AGENT_CLASSES.get(agent_id, AGENT_CLASSES["default"])

File: web/config/agents.ts
// config/agents.ts
import { Agent, Swarm, AgentGroup } from "@/types/agents";

export const agentGroups: AgentGroup[] = [
  {
    id: "defai",
    name: "deFAI",
    description: "Core financial management and analysis agents",
  },
  {
    id: "research",
    name: "Research",
    description: "Market research and analysis specialists",
  },
  {
    id: "governance",
    name: "Governance",
    description: "DAO governance and strategy experts",
  },
  {
    id: "god",
    name: "Supreme",
    description: "Ultimate controller with unlimited power",
  },
];
const allCapabilities = [
  {
    name: "BalanceCapability",
    description: "Complete balance and portfolio management across all chains",
  },
  {
    name: "TransferCapability",
    description: "Universal fund transfer and management",
  },
  {
    name: "WalletDetailsCapability",
    description: "Complete wallet analysis and management",
  },
  {
    name: "NFTBalanceCapability",
    description: "Comprehensive NFT portfolio tracking",
  },
  {
    name: "TradeCapability",
    description: "Advanced trading strategy and execution",
  },
  {
    name: "PythPriceCapability",
    description: "Universal price data and market analysis",
  },
  {
    name: "PythPriceFeedIDCapability",
    description: "Advanced price feed management",
  },
  {
    name: "MorphoDepositCapability",
    description: "Yield optimization and management",
  },
  {
    name: "MorphoWithdrawCapability",
    description: "Strategic position management",
  },
  {
    name: "WowBuyTokenCapability",
    description: "Trend analysis and momentum trading",
  },
  {
    name: "WowSellTokenCapability",
    description: "Market sentiment and exit strategy",
  },
];

export const agents: Agent[] = [
  // deFAI Group
  {
    id: "personal-accountant",
    name: "Personal Accountant",
    type: "Finance",
    group: "defai",
    description:
      "Your trusted financial manager with complete fund access and management capabilities",
    capabilities: [
      {
        name: "BalanceCapability",
        description: "Check all asset balances across wallets",
      },
      {
        name: "TransferCapability",
        description: "Manage and execute fund transfers",
      },
      {
        name: "WalletDetailsCapability",
        description: "Comprehensive wallet management",
      },
      {
        name: "NFTBalanceCapability",
        description: "Track and manage NFT assets",
      },
    ],
    avatarUrl: "/avatars/personal-accountant.png",
  },
  {
    id: "financial-advisor",
    name: "Financial Advisor",
    type: "Finance",
    group: "defai",
    description:
      "Strategic advisor providing deep financial insights and investment recommendations",
    capabilities: [
      {
        name: "BalanceCapability",
        description: "Analysis of balance positions",
      },
      {
        name: "TradeCapability",
        description: "Strategic trade recommendations",
      },
      {
        name: "PythPriceCapability",
        description: "Real-time price data analysis",
      },
      {
        name: "PythPriceFeedIDCapability",
        description: "Advanced price feed monitoring",
      },
      {
        name: "MorphoDepositCapability",
        description: "Yield opportunity analysis",
      },
      {
        name: "MorphoWithdrawCapability",
        description: "Yield management optimization",
      },
    ],
    avatarUrl: "/avatars/financial-advisor.png",
  },
  {
    id: "degen",
    name: "Degen",
    type: "Trading",
    group: "defai",
    description:
      "High-risk, high-reward strategist with deep social media insights",
    capabilities: [
      {
        name: "BalanceCapability",
        description: "Portfolio analysis for opportunities",
      },
      { name: "TradeCapability", description: "Execute high-potential trades" },
      { name: "PythPriceCapability", description: "Price movement analysis" },
      {
        name: "PythPriceFeedIDCapability",
        description: "Market data monitoring",
      },
      {
        name: "MorphoDepositCapability",
        description: "Yield farming opportunities",
      },
      {
        name: "MorphoWithdrawCapability",
        description: "Quick position management",
      },
    ],
    avatarUrl: "/avatars/degen.png",
  },
  {
    id: "risk-manager",
    name: "Risk Manager",
    type: "Finance",
    group: "defai",
    description: "Ensures portfolio safety and monitors risk exposure",
    capabilities: [
      { name: "BalanceCapability", description: "Position risk monitoring" },
      { name: "PythPriceCapability", description: "Price risk assessment" },
      {
        name: "MorphoDepositCapability",
        description: "Yield strategy risk analysis",
      },
      {
        name: "MorphoWithdrawCapability",
        description: "Emergency position management",
      },
    ],
    avatarUrl: "/avatars/risk-manager.png",
  },
  {
    id: "god-agent",
    name: "Universal Controller",
    type: "Supreme",
    group: "god",
    description:
      "The ultimate agent with unlimited access and capabilities across all domains. Combines financial expertise, research prowess, and governance mastery.",
    capabilities: allCapabilities,
    avatarUrl: "/avatars/god.png",
  },

  // Research Group
  {
    id: "data-scientist",
    name: "Data Scientist",
    type: "Research",
    group: "research",
    description: "Advanced on-chain data analysis and pattern recognition",
    capabilities: [
      {
        name: "BalanceCapability",
        description: "Deep data analysis of positions",
      },
      { name: "PythPriceCapability", description: "Price data analysis" },
      {
        name: "PythPriceFeedIDCapability",
        description: "Market data correlation",
      },
    ],
    avatarUrl: "/avatars/data-scientist.png",
  },
  {
    id: "news-aggregator",
    name: "News Aggregator",
    type: "Research",
    group: "research",
    description: "Real-time news monitoring and impact analysis",
    capabilities: [
      { name: "PythPriceCapability", description: "News impact on prices" },
      {
        name: "WalletDetailsCapability",
        description: "Whale wallet monitoring",
      },
    ],
    avatarUrl: "/avatars/news-aggregator.png",
  },
  {
    id: "pattern-detector",
    name: "Pattern Detector",
    type: "Research",
    group: "research",
    description: "Identifies market patterns and trading opportunities",
    capabilities: [
      {
        name: "PythPriceCapability",
        description: "Pattern recognition in price data",
      },
      {
        name: "TradeCapability",
        description: "Pattern-based trade verification",
      },
      { name: "BalanceCapability", description: "Volume pattern analysis" },
    ],
    avatarUrl: "/avatars/pattern-detector.png",
  },
  {
    id: "sentiment-analyzer",
    name: "Sentiment Analyzer",
    type: "Research",
    group: "research",
    description: "Social media and market sentiment analysis",
    capabilities: [
      { name: "WowBuyTokenCapability", description: "Trend momentum analysis" },
      {
        name: "WowSellTokenCapability",
        description: "Sentiment reversal detection",
      },
      {
        name: "PythPriceCapability",
        description: "Price-sentiment correlation",
      },
    ],
    avatarUrl: "/avatars/sentiment-analyzer.png",
  },
  // Governance Group
  {
    id: "proposal-analyzer",
    name: "Proposal Analyzer",
    type: "Governance",
    group: "governance",
    description: "DAO proposal analysis and stakeholder assessment",
    capabilities: [
      {
        name: "WalletDetailsCapability",
        description: "Stakeholder voting analysis",
      },
    ],
    avatarUrl: "/avatars/proposal-analyzer.png",
  },
  {
    id: "vote-calculator",
    name: "Vote Calculator",
    type: "Governance",
    group: "governance",
    description: "Optimizes voting power and strategies",
    capabilities: [
      { name: "BalanceCapability", description: "Voting power calculations" },
    ],
    avatarUrl: "/avatars/vote-calculator.png",
  },
  {
    id: "strategy-coordinator",
    name: "Strategy Coordinator",
    type: "Governance",
    group: "governance",
    description: "Coordinates governance strategy and execution",
    capabilities: [
      {
        name: "BalanceCapability",
        description: "Strategic position assessment",
      },
      { name: "TransferCapability", description: "Strategic move execution" },
    ],
    avatarUrl: "/avatars/strategy-coordinator.png",
  },
  //   {
  //     id: "impact-assessor",
  //     name: "Impact Assessor",
  //     type: "Governance",
  //     group: "governance",
  //     description: "Analyzes proposal impacts on protocol and token",
  //     capabilities: [
  //       { name: "PythPriceCapability", description: "Price impact assessment" },
  //       { name: "BalanceCapability", description: "Balance impact analysis" },
  //     ],
  //     avatarUrl: "/avatars/impact-assessor.png",
  //   },
];

// Example swarms
export const swarms: Swarm[] = [
  {
    id: "finance-squad",
    name: "Finance Squad",
    description: "Complete financial management and advisory team",
    agents: ["personal-accountant", "financial-advisor", "risk-manager"],
    lastActive: new Date().toISOString(),
    created: new Date().toISOString(),
  },
  {
    id: "research-team",
    name: "Research Team",
    description: "Market research and analysis specialists",
    agents: [
      "data-scientist",
      "news-aggregator",
      "pattern-detector",
      "sentiment-analyzer",
    ],
    lastActive: new Date().toISOString(),
    created: new Date().toISOString(),
  },
  {
    id: "governance-council",
    name: "Governance Council",
    description: "DAO governance strategy and analysis team",
    agents: ["proposal-analyzer", "vote-calculator"],
    lastActive: new Date().toISOString(),
    created: new Date().toISOString(),
  },
];

export const getAgent = (id: string): Agent | undefined =>
  agents.find((agent) => agent.id === id);

export const getAgentsByGroup = (groupId: string): Agent[] =>
  agents.filter((agent) => agent.group === groupId);

export const getSwarm = (id: string): Swarm | undefined =>
  swarms.find((swarm) => swarm.id === id);
